File: python/matching-brackets/matching_brackets.py
def is_paired(brackets: str, s: list[str]=[], A='{[(', B='}])'):
    for c in print(brackets) or s.clear() or brackets:
        if c in A: s += c
        elif c in B and (not s or B.index(c) != A.index(s.pop())): return False
    return not s

File: javascript/matching-brackets/matching-brackets.js
// @ts-check

const BRACKETS = '{}[]()', isEven = (n=0) => !(n & 1);

export function isPaired(str='', stack='', idx=-1) { // stack of open brackets
  for (const ch of str)  if ( ~(idx = BRACKETS.indexOf(ch)) ) // found bracket
    if (isEven(idx)) stack += BRACKETS[idx]; // even is open bracket; so add it
    else if (stack.at(-1) == BRACKETS[idx - 1]) stack = stack.slice(0, -1);
    else return false; // if close bracket type isn't last index, it's a fail
  return !stack; } // all bracket pairs match if stack is empty at the end
